IPython/utils/_process_posix.py

```diff
@@ -59,11 +61,18 @@ class ProcessHandler(object):
 
     @property
     def sh(self):
-        if self._sh is None:        
-            self._sh = pexpect.which('sh')
-            if self._sh is None:
-                raise OSError('"sh" shell not found')
-        
+        self._sh = os.environ['SHELL']
```

Comment:
A new change like this can have unexpected consequences, we should either allow users to opt into respecting `$SHELL` or allow them to opt out of this new functionality.